### ISSUE TITLE
 Add public helper method for unwrapping

### ISSUE THREAD

**Issue by vuplea** [ReflectionMagic/ReflectionMagic#31]
We are using ReflectionMagic for a method returning `object` (whose runtime type can be both a primitive or a complex type).
The problem is that `"string".AsDynamic().RealObject` fails because of the wrapper skipping.

I am proposing to expose a helper method for the unwrapping (the code is already present in `PrivateReflectionDynamicObjectBase.Unwrap`).

**Comment by jvandertil** [comment]
Hey, thank you for bringing up this issue and also bringing in a pull request at the same time! Unfortunately I'm having trouble understanding your use case. I do agree that unwrapping whatever `AsDynamic()` is invoked on should have a place somewhere in the library, I just want to make sure we are fixing the right problem.

Per your pull request, you can unblock for now by implementing your own unwrapping code:
```cs
public static object Unwrap(object o)
{
    if (o is PrivateReflectionDynamicObjectBase wrappedObj)
        return wrappedObj.RealObject;

    return o;
}
```

Could you explain why are trying to round-trip the result through a call to `AsDynamic()`? Are you trying to call a method which should be located at runtime based on the real type? Calling `obj.AsDynamic().RealObject` will just give you what you already had: `obj`. Or, as you have noticed, for certain types no wrapping is done, causing that call to throw an exception at runtime.

**Comment by vuplea** [comment]
Thanks for your patience. I understand the use-case might not obvious or common.
We are using ReflectionMagic `.AsDynamic()` in the typical sense, i.e. `instance.AsDynamic().InternalMethod()`.

Things get interesting when the non-public method has a return type of object, but the real type can range from primitive to complex. A real-world example would be: `dynamic deserialized = serializer.AsDynamic().InternalDeserialize(serializedObject)`.

Now, we are not interested in chaining further reflection calls on `deserialized`, but just unwrapping the reflection helper to an `object`. Because we are dealing with compile-time `object`, usages of it will generally involve reflection, e.g. serialization: `JsonConvert.SerializeObject(deserialized)`. The problem would be that the reflection code of the serializer hits the wrapper type instead of the type of the real object.

We've had unwrapping needs in two cases: one for making sure an object is serialized properly, and another in making sure that a proper generic type is created, e.g. `typeof(Generic<>).MakeGenericType(deserialized.GetType())`. This felt like it would warrant creating an issue, in the sense of making `ReflectionMagic` API more complete.

**Comment by jvandertil** [comment]
Sorry for leaving you hanging on this. I'm thinking about an API but can't really get the naming down properly, I came up with this:

```cs
public static class DynamicHelper
{
   public static object Unwrap(dynamic d); 
}
```

I don't like things named `Helper`, but can't really think of anything better.

**Comment by vuplea** [comment]
Since this would be reversing .AsDynamic() which is essence of the ReflectionMagic API, I'm not sure we can come up with a less generic name.
Another not ideal, choice would be `ReflectionMagic` as a class name.

**Comment by jvandertil** [comment]
Fixed in #32
